### PR TITLE
Additions for the chat system

### PR DESCRIPTION
**Pull request by tornac1234**
Some modifications are meant to help for the player list PR (dab3be82461051d2994eb6d37d6f755130648f06), while the rest are just related to the chat system.
Also, mutes aren't persisted at all (reconnecting or restarting server would remove the mute)

- [x] Mute system
- [x] player list complement
- [x] Chat history (#1726)

### CODE REVIEW

File: NitroxServer/Player.cs [hunk 1]
```diff
@@ -33,6 +33,7 @@ public class Player : IProcessorContext
         public NitroxVector3? LastStoredPosition { get; set; }
         public Optional<NitroxId> LastStoredSubRootID { get; set; }
         public ThreadSafeSet<string> CompletedGoals { get; }
+        public bool Muted { get; set; }
```
Comment (dartasen) [inline NitroxServer/Player.cs:36]:
```suggestion
        public bool IsMuted { get; set; }
```

File: NitroxServer/ConsoleCommands/UnmuteCommand.cs [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    {
+        private readonly PlayerManager playerManager;
+
+        public UnmuteCommand(PlayerManager playerManager) : base("unmute", Perms.MODERATOR, "Removes a mute from a player")
+        {
+            this.playerManager = playerManager;
+            AddParameter(new TypePlayer("name", true, "Username to unmute"));
+        }
+
+        protected override void Execute(CallArgs args)
```
Comment (dartasen) [inline NitroxServer/ConsoleCommands/UnmuteCommand.cs:19]: What if the player is not muted ?

File: NitroxServer/ConsoleCommands/UnmuteCommand.cs [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+namespace NitroxServer.ConsoleCommands
+{
+    internal class UnmuteCommand : Command
+    {
+        private readonly PlayerManager playerManager;
+
+        public UnmuteCommand(PlayerManager playerManager) : base("unmute", Perms.MODERATOR, "Removes a mute from a player")
+        {
+            this.playerManager = playerManager;
+            AddParameter(new TypePlayer("name", true, "Username to unmute"));
```
Comment (dartasen) [inline NitroxServer/ConsoleCommands/UnmuteCommand.cs:16]:
```suggestion
            AddParameter(new TypePlayer("name", true, "Player to unmute"));
```

File: NitroxServer/ConsoleCommands/MuteCommand.cs [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    {
+        private readonly PlayerManager playerManager;
+
+        public MuteCommand(PlayerManager playerManager) : base("mute", Perms.MODERATOR, "Prevents a user from chatting")
+        {
+            this.playerManager = playerManager;
+            AddParameter(new TypePlayer("name", true, "Username of the player to mute"));
+        }
+
+        protected override void Execute(CallArgs args)
```
Comment (dartasen) [inline NitroxServer/ConsoleCommands/MuteCommand.cs:19]: What if a player is already muted ?

File: NitroxServer/ConsoleCommands/MuteCommand.cs [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+namespace NitroxServer.ConsoleCommands
+{
+    internal class MuteCommand : Command
+    {
+        private readonly PlayerManager playerManager;
+
+        public MuteCommand(PlayerManager playerManager) : base("mute", Perms.MODERATOR, "Prevents a user from chatting")
+        {
+            this.playerManager = playerManager;
+            AddParameter(new TypePlayer("name", true, "Username of the player to mute"));
```
Comment (dartasen) [inline NitroxServer/ConsoleCommands/MuteCommand.cs:16]:
```suggestion
            AddParameter(new TypePlayer("name", true, "Player to mute"));
```

File: NitroxServer/ConsoleCommands/DeopCommand.cs [hunk 6]
```diff
@@ -16,6 +17,9 @@ protected override void Execute(CallArgs args)
             Player targetPlayer = args.Get<Player>(0);
             targetPlayer.Permissions = Perms.PLAYER;
 
+            // Need to notify him so that he no longer shows admin stuff on client (which would in any way stop working)
+            targetPlayer.SendPacket(new OpPlayer(false));
```
Comment (dartasen) [inline NitroxServer/ConsoleCommands/DeopCommand.cs:21]: Don't forget that there is a promote command

File: NitroxServer/ConsoleCommands/Abstract/Command.cs [hunk 7]
```diff
@@ -131,7 +131,7 @@ public static void SendMessageToPlayer(Optional<Player> player, string message)
         public static void SendMessage(Optional<Player> player, string message)
         {
             SendMessageToPlayer(player, message);
-            Log.Info(message);
+            Log.Info($"{(player.HasValue ? $"=> {player.Value.Name}: ": "")}{message}");
```
Comment (dartasen) [inline NitroxServer/ConsoleCommands/Abstract/Command.cs:134]: ?

File: NitroxServer/Communication/Packets/Processors/ChatMessageProcessor.cs [hunk 8]
```diff
@@ -15,8 +15,13 @@ public ChatMessageProcessor(PlayerManager playerManager)
...
         {
-            Log.Info(string.Format("<{0}>: {1}", player.Name, packet.Text));
-            playerManager.SendPacketToOtherPlayers(packet, player);
+            if (player.Muted)
+            {
+                player.SendPacket(new ChatMessage(ChatMessage.SERVER_ID, "You're currently muted"));
+                return;
+            }
+            Log.Info($"<{player.Name}>: {packet.Text}");
+            playerManager.SendPacketToAllPlayers(packet);
```
Comment (dartasen) [inline NitroxServer/Communication/Packets/Processors/ChatMessageProcessor.cs:24]:
```suggestion
            playerManager.SendPacketToOtherPlayers(packet, player);
```

Comment (Jannify) [reply NitroxServer/Communication/Packets/Processors/ChatMessageProcessor.cs:24]: I think this is intentional as this is the only way for the client to know if he is (still) muted

Comment (tornac1234) [reply NitroxServer/Communication/Packets/Processors/ChatMessageProcessor.cs:24]: yes this is to lend the authority to the server

File: NitroxServer/ConsoleCommands/Abstract/Command.cs (hunk 7 above)
Comment (tornac1234) [reply NitroxServer/ConsoleCommands/Abstract/Command.cs:134]: this logging was weird before so this lets you understand which message is sent to which player or to the server

**Review by Jannify — CHANGES_REQUESTED** [review]

File: NitroxClient/MonoBehaviours/Gui/Chat/PlayerChatInputField.cs [hunk 9]
```diff
@@ -45,23 +80,67 @@ private void Update()
                 return;
             }
 
-            if (selected && InputField.text != "")
+            if (selected)
             {
-                ResetTimer();
-                if (UnityEngine.Input.GetKey(KeyCode.Return))
+                if (InputField.text != "")
```
Comment (Jannify) [inline NitroxClient/MonoBehaviours/Gui/Chat/PlayerChatInputField.cs:85]:
```suggestion
                if (string.IsNullOrEmpty(InputField.text))
```

**Review by Jannify — APPROVED** [review]
Seems fine code wise besides the one suggested change.

**Review by dartasen — APPROVED** [review]

File: NitroxServer/ConsoleCommands/Abstract/Command.cs [hunk 10]
```diff
@@ -131,7 +130,14 @@ public static void SendMessageToPlayer(Optional<Player> player, string message)
...
             SendMessageToPlayer(player, message);
-            Log.Info(message);
+            if (player.HasValue)
+            {
+                Log.Info($"=> {player.Value.Name}: {message}");
+            }
+            else
+            {
+                Log.Info(message);
+            }
```
Comment (dartasen) [inline NitroxServer/ConsoleCommands/Abstract/Command.cs:140]: Imo, we don't need to see what player are seeing

File: NitroxServer/ConsoleCommands/Abstract/Command.cs [hunk 11]
```diff
@@ -141,7 +147,7 @@ public static void SendMessageToAllPlayers(string message)
         {
             PlayerManager playerManager = NitroxServiceLocator.LocateService<PlayerManager>();
             playerManager.SendPacketToAllPlayers(new ChatMessage(ChatMessage.SERVER_ID, message));
-            Log.Info(message);
+            Log.Info($"=> everyone: {message}");
```
Comment (dartasen) [inline NitroxServer/ConsoleCommands/Abstract/Command.cs:150]:
```suggestion
            Log.Info($"[BROADCAST] {message}");
```

File: Nitrox.Assets.Subnautica/LanguageFiles/English.json [hunk 12]
```diff
@@ -38,7 +38,7 @@
     "Nitrox_WaitingPassword": "Waiting for server password input…",
     "Nitrox_WaitingUserInput": "Waiting for user input…",
     "Nitrox_LaunchGame": "Launching game…",
-    "Nitrox_SilencedChatNotif": "Chat is now muted",
```
Comment (Jannify) [inline Nitrox.Assets.Subnautica/LanguageFiles/English.json:41]: I would like to move all language file changes to a separate pr as I fear a merge conflict with weblate. Therefor this new pr will also be merged after the weblate pr, meaning shortly before the next release.

**Review by Jannify — CHANGES_REQUESTED** [review]

**Comment by Jannify** [comment]
In-game test with Ninja and Hen was successful. Waiting for applying suggested changes and (probably) needed rebase.

File: NitroxClient/MonoBehaviours/Gui/Chat/PlayerChat.cs [hunk 13]
```diff
@@ -50,13 +50,14 @@ public IEnumerator SetupChatComponents()
...
-            backgroundImages = new[] { transform.GetChild(0).GetComponent<Image>(), transform.GetChild(1).GetComponent<Image>(), transform.GetChild(3).GetComponent<Image>() };
+            // We pick any image that's inside the chat component to have all of their opacity lowered
+            backgroundImages = transform.GetComponentsInChildren<Image>();
 
             yield return new WaitForEndOfFrame(); //Needed so Select() works on initialization
             IsReady = true;
             if (NitroxPrefs.SilenceChat.Value)
             {
-                Log.InGame(Language.main.Get("Nitrox_SilencedChatNotif"));
+                Log.InGame(Language.main.Get("Nitrox_SilencedChatNotify"));
```
Comment (Jannify) [inline NitroxClient/MonoBehaviours/Gui/Chat/PlayerChat.cs:60]: As we delay the key change after weblate merge this should also be reverted

**Review by Jannify — CHANGES_REQUESTED** [review]

File: NitroxClient/MonoBehaviours/Gui/Chat/PlayerChat.cs (hunk 13 above)
Comment (tornac1234) [reply NitroxClient/MonoBehaviours/Gui/Chat/PlayerChat.cs:60]: Ah yes well found

**Review by dartasen — APPROVED** [review]
